compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -323,17 +345,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
         }
-
-        let mut err = struct_span_err!(
-            tcx.sess,
-            span,
-            E0599,
-            "no {} named `{}` found for {} `{}` in the current scope",
-            item_kind,
-            item_name,
-            rcvr_ty.prefix_string(self.tcx),
-            ty_str_reported,
-        );
+        let mut err;
+        if sugg_span.ctxt().outer_expn_data().macro_def_id.map_or(false, |def_id| {
+            tcx.is_diagnostic_item(sym::write_macro, def_id)
+                || tcx.is_diagnostic_item(sym::writeln_macro, def_id)
+        }) && item_name.name == Symbol::intern("write_fmt")
+            && args.is_some()
+        {
+            err = self.suggest_missing_writer(rcvr_ty, args).unwrap();
+        } else {
+            err = struct_span_err!(
+                tcx.sess,
+                span,
+                E0599,
+                "no {} named `{}` found for {} `{}` in the current scope",
+                item_kind,
+                item_name,
+                rcvr_ty.prefix_string(self.tcx),
+                ty_str_reported,
+            );
+        }
```

Comment:
Several nits:
1. No need to unwrap here
2. No need to create a binding then assign it from both branches

```suggestion
        let mut err = if sugg_span.ctxt().outer_expn_data().macro_def_id.map_or(false, |def_id| {
            tcx.is_diagnostic_item(sym::write_macro, def_id)
                || tcx.is_diagnostic_item(sym::writeln_macro, def_id)
        }) && item_name.name == Symbol::intern("write_fmt")
            && let Some(args) = args
        {
            self.suggest_missing_writer(rcvr_ty, args)
        } else {
            struct_span_err!(
                tcx.sess,
                span,
                E0599,
                "no {} named `{}` found for {} `{}` in the current scope",
                item_kind,
                item_name,
                rcvr_ty.prefix_string(self.tcx),
                ty_str_reported,
            )
        }
```

Then you'll need to adjust the signature of `suggest_missing_writer` to no longer take or return an `Option`.